cdc/cdc.cc

```diff
@@ -323,6 +329,75 @@ class transformer final {
         }
         return res;
     }
+
+    static db::timeout_clock::time_point default_timeout() {
+        return db::timeout_clock::now() + 10s;
+    }
+
+    future<lw_shared_ptr<cql3::untyped_result_set>> pre_image_select(
+            service::storage_proxy& proxy,
+            service::client_state& client_state,
+            db::consistency_level cl,
+            const mutation& m)
+    {
+        auto& p = m.partition();
+        if (p.partition_tombstone() || !p.row_tombstones().empty() || p.clustered_rows().empty()) {
+            return make_ready_future<lw_shared_ptr<cql3::untyped_result_set>>();
+        }
+
+        auto& pk = m.key();
+
+        dht::partition_range_vector partition_ranges{dht::partition_range(dht::global_partitioner().decorate_key(*_schema, pk))};
+
+        auto&& pc = _schema->partition_key_columns();
+        auto&& cc = _schema->clustering_key_columns();
+
+        std::vector<query::clustering_range> bounds;
+        if (cc.empty()) {
+            bounds.push_back(query::clustering_range::make_open_ended_both_sides());
+        } else {
+            for (const rows_entry& r : p.clustered_rows()) {
+                auto& ck = r.key();
+                bounds.push_back(query::clustering_range::make_singular(ck));
+            }
+        }
+
+        std::vector<const column_definition*> columns;
+        columns.reserve(_schema->all_columns().size());
+
+        std::transform(pc.begin(), pc.end(), std::back_inserter(columns), [](auto& c) { return &c; });
+        std::transform(cc.begin(), cc.end(), std::back_inserter(columns), [](auto& c) { return &c; });
+
+        query::column_id_vector static_columns, regular_columns;
+
+        auto* cids = &static_columns;
+        auto kind = column_kind::static_column;
+
+        for (const row& r : { std::ref(p.static_row()), std::ref(p.clustered_rows().begin()->row().cells()) }) {
+            r.for_each_cell([&](column_id id, const atomic_cell_or_collection&) {
+                auto& cdef =_schema->column_at(kind, id);
+                cids->emplace_back(id);
+                columns.emplace_back(&cdef);
+            });
+            cids = &regular_columns;
```

Comment:
Duplicated code is the devil. Updating a var in a loop is not strange. Would be nicer to make the secondary var a part of the loop data, but since it is just two, this is simpler.

Reply:
or just use transform with back inserter.

Reply:
@elcallio the funny thing that the "unduplicated" code is actually longer than the duplicated code, and harder to understand (it took me a minute to figure out why you're overwriting these variables at the end of each iteration - and that it's only meant to happen after the first iteration). Its only merit is that it doesn't have the same line of code twice. Well, sort of (you do have cids= assigned in two places). I guess it's a matter of taste.